readthedocs/search/api/v3/views.py

```diff
@@ -0,0 +1,134 @@
+
+    def _validate_query_params(self):
+        query = self.request.GET.get("q")
+        errors = {}
+        if not query:
+            errors["q"] = [_("This query parameter is required")]
+        if errors:
+            raise ValidationError(errors)
+
+    @cached_property
+    def _backend(self):
+        backend = self.search_backend_class(
+            request=self.request,
+            query=self.request.GET["q"],
+        )
+        return backend
+
+    def _get_search_query(self):
+        return self._backend.parser.query
+
+    def _get_projects_to_search(self):
+        return self._backend.projects
+
+    def get_queryset(self):
+        """
+        Returns an Elasticsearch DSL search object or an iterator.
+
+        .. note::
+
+           Calling ``list(search)`` over an DSL search object is the same as
+           calling ``search.execute().hits``. This is why an DSL search object
+           is compatible with DRF's paginator.
+        """
+        use_advanced_query = should_use_advanced_query(self._get_projects_to_search())
+        search = self._backend.search(
+            use_advanced_query=use_advanced_query,
+            aggregate_results=False,
+        )
+        if not search:
+            return []
+
+        return search
+
+    def get(self, request, *args, **kwargs):
+        self._validate_query_params()
+        result = self.list()
+        self._record_query(result)
+        return result
+
+    def _record_query(self, response):
+        total_results = response.data.get("count", 0)
+        time = timezone.now()
+        query = self._get_search_query().lower().strip()
+        # NOTE: I think this may be confusing,
+        # since the number of results is the total
+        # of searching on all projects, this specific project
+        # could have had 0 results.
+        for project, version in self._get_projects_to_search():
+            tasks.record_search_query.delay(
+                project.slug,
+                version.slug,
+                query,
+                total_results,
+                time.isoformat(),
+            )
+
+    def list(self):
+        queryset = self.get_queryset()
+        page = self.paginator.paginate_queryset(
+            queryset,
+            self.request,
+            view=self,
+        )
+        serializer = self.get_serializer(
+            page, many=True, projects=self._get_projects_to_search()
+        )
+        response = self.paginator.get_paginated_response(serializer.data)
+        self._add_extra_fields(response)
+        return response
+
+    def _add_extra_fields(self, response):
+        """Add additional fields to the response."""
+        response.data["projects"] = [
+            {"slug": project.slug, "versions": [{"slug": version.slug}]}
+            for project, version in self._get_projects_to_search()
+        ]
+        response.data["query"] = self._get_search_query()
+
+
+class BaseProxiedSearchAPI(SearchAPI):
```

Comment:
This should explain why it's here since it seems odd without an explanation -- I assume for some downstream reason?